discovery/ovhcloud/ovhcloud.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ovhcloud
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"inet.af/netaddr"
+
+	"github.com/fatih/structs"
+	"github.com/go-kit/log"
+	"github.com/go-playground/validator/v10"
+	"github.com/grafana/regexp"
+	"github.com/ovh/go-ovh/ovh"
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/discovery/refresh"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+)
+
+// metaLabelPrefix is the meta prefix used for all meta labels.
+// in this discovery.
+const (
+	metaLabelPrefix = model.MetaLabelPrefix + "ovhcloud_"
+)
+
+func addFieldsOnLabels(fields []*structs.Field, labels model.LabelSet, prefix string) {
+	for _, f := range fields {
+		labelName := f.Tag("label")
+		if labelName == "-" {
+			// Skip labels with -
+			continue
+		}
+		if labelName == "" {
+			labelName = f.Tag("json")
+		}
+
+		if labelName != "" {
+			labels[model.LabelName(prefix+labelName)] = model.LabelValue(fmt.Sprintf("%+v", f.Value()))
+		}
+	}
+}
+
+type refresher interface {
+	refresh(context.Context) ([]*targetgroup.Group, error)
+	getSource() string
+	getService() string
+}
+
+// AuthDetails partial
+type AuthDetails struct {
+	User string `json:"user"`
+}
+
+// SDConfig sd config
+type SDConfig struct {
+	Endpoint          string         `yaml:"endpoint" validate:"required"`
+	ApplicationKey    string         `yaml:"application_key" validate:"required"`
+	ApplicationSecret string         `yaml:"application_secret" validate:"required"`
+	ConsumerKey       string         `yaml:"consumer_key" validate:"required"`
+	RefreshInterval   model.Duration `yaml:"refresh_interval" validate:"required"`
+	Service           string         `yaml:"service" validate:"required,oneof=dedicated_server vps"`
+	NoAuthTest        bool           `yaml:"no_auth_test"`
+}
+
+// IPs struct to store IPV4 and IPV6
+type IPs struct {
+	IPV4 string `json:"ipv4" label:"ipv4"`
+	IPV6 string `json:"ipv6" label:"ipv6"`
+}
+
+// Name get name
+func (c SDConfig) Name() string {
+	return "ovhcloud"
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+
+	validate := validator.New()
```

Comment:
Why is this needed?

Reply:
Is to check validate tag on struct (required, oneof [validator](https://github.com/go-playground/validator))
https://github.com/prometheus/prometheus/pull/10802/commits/12f18c2772a5a9a2499877572c76bf77628c49d8#diff-643d27d894488ad9f1633a2542dd9fe466432bf97880359a5eaab2b9ae8ea4e7R78

Like this 
https://github.com/prometheus/prometheus/pull/10802/commits/4767a7e265c845efaf3fe8fcdeccedaf3062052a#diff-643d27d894488ad9f1633a2542dd9fe466432bf97880359a5eaab2b9ae8ea4e7R77
To do this kind of test with a validate tag
https://github.com/prometheus/prometheus/blob/main/discovery/scaleway/scaleway.go#L65